web/pdf_find_controller.js

```diff
@@ -331,193 +462,140 @@ class PDFFindController {
         return false;
       }
     }
-    const endIdx = startIdx + length - 1;
-    if (endIdx < content.length - 1) {
-      const last = content.charCodeAt(endIdx);
-      const limit = content.charCodeAt(endIdx + 1);
+
+    match = content
+      .slice(startIdx + length)
+      .match(NOT_DIACRITIC_FROM_START_REG_EXP);
+    if (match) {
+      const last = content.charCodeAt(startIdx + length - 1);
+      const limit = match[1].charCodeAt(0);
       if (getCharacterType(last) === getCharacterType(limit)) {
         return false;
       }
     }
+
     return true;
   }
 
-  _calculatePhraseMatch(query, pageIndex, pageContent, pageDiffs, entireWord) {
+  _calculateRegExpMatch(query, entireWord, pageIndex, pageContent) {
     const matches = [],
       matchesLength = [];
-    const queryLen = query.length;
 
-    let matchIdx = -queryLen;
-    while (true) {
-      matchIdx = pageContent.indexOf(query, matchIdx + queryLen);
-      if (matchIdx === -1) {
-        break;
-      }
-      if (entireWord && !this._isEntireWord(pageContent, matchIdx, queryLen)) {
+    const diffs = this._pageDiffs[pageIndex];
+    let match;
+    while ((match = query.exec(pageContent)) !== null) {
+      if (
+        entireWord &&
+        !this._isEntireWord(pageContent, match.index, match[0].length)
+      ) {
         continue;
       }
-      const originalMatchIdx = getOriginalIndex(matchIdx, pageDiffs),
-        matchEnd = matchIdx + queryLen - 1,
-        originalQueryLen =
-          getOriginalIndex(matchEnd, pageDiffs) - originalMatchIdx + 1;
 
-      matches.push(originalMatchIdx);
-      matchesLength.push(originalQueryLen);
+      const [matchPos, matchLen] = getOriginalIndex(
+        diffs,
+        match.index,
+        match[0].length
+      );
+      matches.push(matchPos);
+      matchesLength.push(matchLen);
     }
     this._pageMatches[pageIndex] = matches;
     this._pageMatchesLength[pageIndex] = matchesLength;
   }
 
-  _calculateWordMatch(query, pageIndex, pageContent, pageDiffs, entireWord) {
-    const matchesWithLength = [];
-
-    // Divide the query into pieces and search for text in each piece.
-    const queryArray = query.match(/\S+/g);
-    for (let i = 0, len = queryArray.length; i < len; i++) {
-      const subquery = queryArray[i];
-      const subqueryLen = subquery.length;
-
-      let matchIdx = -subqueryLen;
-      while (true) {
-        matchIdx = pageContent.indexOf(subquery, matchIdx + subqueryLen);
-        if (matchIdx === -1) {
-          break;
+  _convertToRegExpString(query, hasDiacritics) {
+    const { matchDiacritics } = this._state;
+    query = query.replace(
+      SPECIAL_CHARS_REG_EXP,
+      (
+        match,
+        p1 /* to escape */,
+        p2 /* punctuation */,
+        p3 /* whitespaces */,
+        p4 /* diacritics */,
+        p5 /* letters */
+      ) => {
+        if (p1) {
+          // Escape characters like *+?... to not interfer with regexp syntax.
+          return `\\s*\\${p1}\\s*`;
```

Comment:
*This is potentially a case of a useless micro-optimization, but I wanted to ask nonetheless:*
Given that we call `getTextContent` with the `normalizeWhitespace = true` option set, all whitespaces are thus normal spaces (i.e. 0x20), hence I wonder if it would perhaps be more efficient to only check for that character rather than arbitrary white-space in these regular expressions?